### PR TITLE
chore: Add conversion from `v1beta1/NodeClass` to `v1alpha1/AWSNodeTemplate`

### PR DESCRIPTION
**Pull request by jonathan-innis**
Fixes #N/A 

**Description**

Adds conversion from a `v1beta1/NodeClass` to a `v1alpha5/AWSNodeTemplate` for the controller and hash annotation handling. This allows the business logic for the controllers and the CloudProvider to only be written with the `v1beta1` APIs in mind.

**How was this change tested?**

`make presubmit`

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates 
- [ ] Yes, issue opened: # 
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/test/expectations/expectations.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+	. "github.com/onsi/gomega" //nolint:revive,stylecheck
+	"github.com/samber/lo"
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/aws/karpenter/pkg/apis/v1alpha1"
+	"github.com/aws/karpenter/pkg/apis/v1beta1"
+)
+
+func ExpectBlockDeviceMappingsEqual(bdm1 []*v1alpha1.BlockDeviceMapping, bdm2 []*v1beta1.BlockDeviceMapping) {
```
Comment (njtran) [inline pkg/test/expectations/expectations.go:28]: Just curious, should we be asserting anything about the ordering of this? Don't we consider root volumes/default volume size by the ordering in this list? Should we care about that?

File: pkg/utils/nodeclass/nodeclass.go [hunk 2]
```diff
@@ -36,66 +37,79 @@ func New(nodeTemplate *v1alpha1.AWSNodeTemplate) *v1beta1.NodeClass {
...
+			Subnets:        NewSubnets(nodeTemplate.Status.Subnets),
+			SecurityGroups: NewSecurityGroups(nodeTemplate.Status.SecurityGroups),
+			AMIs:           NewAMIs(nodeTemplate.Status.AMIs),
 		},
 		IsNodeTemplate: true,
 	}
 }
 
-func NewSubnetSelectorTerms(subnetSelector map[string]string) []v1beta1.SubnetSelectorTerm {
+func NewSubnetSelectorTerms(subnetSelector map[string]string) (terms []v1beta1.SubnetSelectorTerm) {
```
Comment (njtran) [inline pkg/utils/nodeclass/nodeclass.go:65]: nit: named return values

File: pkg/utils/nodeclass/nodeclass.go [hunk 3]
```diff
@@ -36,66 +37,79 @@ func New(nodeTemplate *v1alpha1.AWSNodeTemplate) *v1beta1.NodeClass {
...
+		terms = append(terms, v1beta1.SubnetSelectorTerm{
+			Tags: tags,
+			ID:   id,
+		})
 	}
+	return terms
 }
 
-func NewSecurityGroupSelectorTerms(securityGroupSelector map[string]string) []v1beta1.SecurityGroupSelectorTerm {
+func NewSecurityGroupSelectorTerms(securityGroupSelector map[string]string) (terms []v1beta1.SecurityGroupSelectorTerm) {
```
Comment (njtran) [inline pkg/utils/nodeclass/nodeclass.go:90]: Same here

File: pkg/utils/nodeclass/nodeclass.go [hunk 4]
```diff
@@ -173,6 +196,43 @@ func NewMetadataOptions(mo *v1alpha1.MetadataOptions) *v1beta1.MetadataOptions {
 	}
 }
 
+func NewSubnets(subnets []v1alpha1.Subnet) []v1beta1.Subnet {
+	if subnets == nil {
```
Comment (njtran) [inline pkg/utils/nodeclass/nodeclass.go:200]: Just sanity checking that lo.Map() doesn't take in a nil list?

File: pkg/test/expectations/expectations.go (hunk 1 above)
Comment (jonathan-innis) [reply pkg/test/expectations/expectations.go:28]: Ah, that's a good point. Yes, we should be asserting the order here.

File: pkg/utils/nodeclass/nodeclass.go (hunk 3 above)
Comment (jonathan-innis) [reply pkg/utils/nodeclass/nodeclass.go:90]: I generally take the approach of naming the return parameters when the return value is a list and we are going to have to declare it in the body of the function anyways with a `var` statement

File: pkg/utils/nodeclass/nodeclass.go (hunk 2 above)
Comment (jonathan-innis) [reply pkg/utils/nodeclass/nodeclass.go:65]: I generally take the approach of naming the return parameters when the return value is a list and we are going to have to declare it in the body of the function anyways with a `var` statement so I personally think that this is ok here.

File: pkg/utils/nodeclass/nodeclass.go (hunk 4 above)
Comment (njtran) [reply pkg/utils/nodeclass/nodeclass.go:200]: Ahhh makes sense. Thanks!

**Review by njtran — APPROVED** [review]

**Pull request merged**